tests/test_tfe.py

```diff
@@ -40,116 +36,129 @@
 def expZ(state):
     return np.abs(state[0]) ** 2 - np.abs(state[1]) ** 2
 
+@pytest.fixture()
+def skip_if_no_tf_support(tf_support):
+    if not tf_support:
+        pytest.skip("Skipped, no tfe support")   
 
-class TFEQNodeTests(BaseTest):
-    """TFEQNode basic tests."""
-    def setUp(self):
-        if not tf_support:
-            self.skipTest('TFE interface not tested')
-
-        self.dev1 = qml.device('default.qubit', wires=1)
-        self.dev2 = qml.device('default.qubit', wires=2)
-
-    def test_qnode_fail(self):
-        """Tests that QNode initialization failures correctly raise exceptions."""
-        self.logTestName()
-        par = tfe.Variable(0.5)
 
-        #---------------------------------------------------------
-        ## faulty quantum functions
+@pytest.mark.usefixtures("skip_if_no_tf_support")
+class TestTFEQNodeExceptions():
+    """TFEQNode basic tests."""
 
-        # qfunc must return only Expectations
-        @qml.qnode(self.dev2, interface='tfe')
+    def test_qnode_fails_on_wrong_return_type(self, qubit_device_2_wires):
+        """The qfunc must return only Expectations"""
+        @qml.qnode(qubit_device_2_wires, interface='tfe')
         def qf(x):
             qml.RX(x, wires=[0])
             return qml.expval(qml.PauliZ(0)), 0.3
 
-        with self.assertRaisesRegex(QuantumFunctionError, 'must return either'):
-            qf(par)
+        with pytest.raises(QuantumFunctionError, match='must return either'):
+            qf(tfe.Variable(0.5))
+
+    def test_qnode_fails_on_expval_not_returned(self, qubit_device_2_wires):
+        """All expectation values in the qfunc must be returned"""
 
-        # all EVs must be returned...
-        @qml.qnode(self.dev2, interface='tfe')
+        @qml.qnode(qubit_device_2_wires, interface='tfe')
         def qf(x):
             qml.RX(x, wires=[0])
             ex = qml.expval(qml.PauliZ(1))
             return qml.expval(qml.PauliZ(0))
 
-        with self.assertRaisesRegex(QuantumFunctionError, 'All measured observables'):
-            qf(par)
+        with pytest.raises(QuantumFunctionError, match='All measured observables'):
+            qf(tfe.Variable(0.5))
 
-        # ...in the correct order
-        @qml.qnode(self.dev2, interface='tfe')
+    def test_qnode_fails_on_wrong_expval_order(self, qubit_device_2_wires):
+        """Expvals must be returned in the order they were created in"""
+
+        @qml.qnode(qubit_device_2_wires, interface='tfe')
         def qf(x):
             qml.RX(x, wires=[0])
             ex = qml.expval(qml.PauliZ(1))
             return qml.expval(qml.PauliZ(0)), ex
 
-        with self.assertRaisesRegex(QuantumFunctionError, 'All measured observables'):
-            qf(par)
+        with pytest.raises(QuantumFunctionError, match='All measured observables'):
```

Comment:
This match is the same as the previous one, which may mask different failure cases. @johannesjmeyer do you know if the full text is different in these two cases?

Reply:
It isn't. They both come from these lines of code:
```python
# check that all ev's are returned, in the correct order
if res != tuple(self.ev):
    raise QuantumFunctionError("All measured observables must be returned in the "
                               "order they are measured.")
```
which checks both conditions at the same time.